core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,147 @@
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename T>
+class HostSharedPtr {
+ public:
+  template <class Deleter>
+  HostSharedPtr(T* value_ptr, bool owning, Deleter deleter)
+      : m_value_ptr(value_ptr),
+        m_deleter(owning ? (new std::function<void(T*)>(std::move(deleter)))
+                         : nullptr),
+        m_counter(owning ? (new int(1)) : nullptr) {}
+
+  KOKKOS_FUNCTION HostSharedPtr(HostSharedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr),
+        m_deleter(other.m_deleter),
+        m_counter(other.m_counter) {
+    other.m_value_ptr = nullptr;
+    other.m_deleter   = nullptr;
+    other.m_counter   = nullptr;
+  }
+
+  KOKKOS_FUNCTION HostSharedPtr(const HostSharedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr),
+        m_deleter(other.m_deleter),
+        m_counter(other.m_counter) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_counter) Kokkos::atomic_add(m_counter, 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION HostSharedPtr& operator=(HostSharedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_deleter         = other.m_deleter;
+      other.m_deleter   = nullptr;
+      m_counter         = other.m_counter;
+      other.m_counter   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION HostSharedPtr& operator=(
+      const HostSharedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_deleter   = other.m_deleter;
+      m_counter   = other.m_counter;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (m_counter) Kokkos::atomic_add(m_counter, 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~HostSharedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept { return *get(); }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the HostSharedPtr manages the lifetime of the object
+  KOKKOS_FUNCTION bool owner() const noexcept { return m_counter != nullptr; }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (m_counter == nullptr) return;
+    int const count = Kokkos::atomic_fetch_sub(m_counter, 1);
+    if (count == 1) {
+      delete m_counter;
```

Comment:
Don't you need to set `m_counter` to `nullptr`? The value of `m_counter` is undefined and it could be equal to two if you are unlucky and if `cleanup()` is called again, you would try to free memory that has already been freed.

Reply:
All places where we call this private function set `m_counter` later (or we are in the destructor) so that should not happen. It might be a good idea to make sure anyway. I added a commit setting all the pointers to `nullptr`.